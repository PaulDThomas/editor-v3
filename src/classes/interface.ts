import { EditorV3AtBlock, IEditorV3AtBlock } from "./EditorV3AtBlock";
import { IEditorV3SelectBlock } from "./EditorV3SelectBlock";
import { EditorV3TextBlock, IEditorV3TextBlock } from "./EditorV3TextBlock";
import { IMarkdownSettings } from "./markdown/MarkdownSettings";

/**
 * Available text alignments
 */
export enum EditorV3Align {
  "left" = "left",
  "center" = "center",
  "decimal" = "decimal",
  "right" = "right",
}

/**
 * Object container for all avaialble styles, name: React.CSSProperties
 */
export interface EditorV3Style extends React.CSSProperties {
  isLocked?: boolean;
  isNotAvailabe?: boolean;
}
export interface EditorV3Styles extends Record<string, EditorV3Style> {
  [styleName: string]: EditorV3Style;
}

/**
 * Line object for EditorV3
 */
export interface EditorV3LineProps {
  textBlocks: EditorV3TextBlock[];
}

export interface EditorV3DropListItem<T extends Record<string, string>> {
  text: string;
  data?: T;
  listRender?: HTMLLIElement;
}

/**
 * Text block class objects EditorV3
 */
export type EditorV3BlockClass = EditorV3TextBlock | EditorV3AtBlock;

/**
 * Content object for EditorV3
 * @param textAlignment Alignment of text withing the editor
 * @param decimalAlignPercent Percent of width to align decimal to, 0-100;
 * @param caret Current position of the caret
 * @param styles Available styles
 * @param showMarkdown Print markdown syntax instead of rendering
 * @param markdownSettings Settings for markdown
 */
export interface EditorV3ContentProps {
  allowMarkdown: boolean;
  allowNewLine: boolean;
  allowWindowView: boolean;
  decimalAlignPercent: number;
  markdownSettings: IMarkdownSettings;
  showMarkdown: boolean;
  styles?: EditorV3Styles;
  textAlignment: EditorV3Align;
  atListFunction?: (typedString: string) => Promise<EditorV3DropListItem<Record<string, string>>[]>;
  maxAtListLength: number;
}
export interface EditorV3ContentPropsInput {
  allowMarkdown?: boolean;
  allowNewLine?: boolean;
  allowWindowView?: boolean;
  decimalAlignPercent?: number;
  markdownSettings?: IMarkdownSettings;
  showMarkdown?: boolean;
  styles?: EditorV3Styles;
  textAlignment?: EditorV3Align;
  atListFunction?: (typedString: string) => Promise<EditorV3DropListItem<Record<string, string>>[]>;
  maxAtListLength?: number;
}

/**
 * Line import object for EditorV3
 */
export interface IEditorV3Line {
  textBlocks: (IEditorV3AtBlock | IEditorV3TextBlock | IEditorV3SelectBlock)[];
  contentProps?: EditorV3ContentPropsInput;
}

/**
 * Complete import object for EditorV3
 */
export interface IEditorV3 {
  lines: IEditorV3Line[];
  contentProps?: EditorV3ContentPropsInput;
}

/**
 * Cursor position object for EditorV3
 */
export interface EditorV3Position {
  startLine: number;
  startChar: number;
  endLine: number;
  endChar: number;
  isCollapsed?: boolean;
  focusAt?: "start" | "end";
}
export interface EditorV3PositionF {
  initialLine: number;
  initialChar: number;
  focusLine: number;
  focusChar: number;
}
/**
 * Word positions in the content
 */
export interface EditorV3WordPosition {
  line: number;
  startChar: number;
  endChar: number;
  isLocked: boolean;
}

/**
 * Render properties passed down to elements
 */
export interface EditorV3RenderProps {
  atListFunction?: (typedString: string) => Promise<EditorV3DropListItem<Record<string, string>>[]>;
  caretPosition?: EditorV3Position | null;
  currentEl?: HTMLDivElement | HTMLSpanElement;
  doNotSplitWordSpans?: boolean;
  editableEl?: HTMLDivElement;
  markdownSettings?: IMarkdownSettings;
  maxAtListLength?: number;
  styles?: EditorV3Styles;
}
