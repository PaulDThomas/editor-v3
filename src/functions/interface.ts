// Interfaces
export interface iColourStyles {
  [styleName: string]: React.CSSProperties;
}

export interface iStyleBlock {
  start: number;
  end: number;
  styleName?: string;
}

export interface iColouredLine {
  text: string;
  styles?: iColourStyles;
  styleBlocks?: iStyleBlock[];
}

export enum EditorV3Align {
  'left' = 'left',
  'center' = 'center',
  'decimal' = 'decimal',
  'right' = 'right',
}

export interface AieStyleMap {
  [styleName: string]: { css: React.CSSProperties; aieExclude: string[] };
}
export interface AieStyleExcludeMap {
  [styleName: string]: string[];
}
