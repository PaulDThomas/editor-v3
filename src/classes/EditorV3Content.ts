import { EditorV3Line } from "./EditorV3Line";
import { EditorV3TextBlock } from "./EditorV3TextBlock";
import {
  EditorV3Align,
  EditorV3ContentProps,
  EditorV3Import,
  EditorV3Position,
  EditorV3Styles,
} from "./interface";
import { IMarkdownSettings, defaultMarkdownSettings } from "./markdown/MarkdownSettings";
import { readV3Html } from "../functions/readV3Html";
import { MarkdownLineClass } from "./markdown/MarkdownLineClass";

/**
 * Represents the content of an EditorV3 instance.
 */
export class EditorV3Content {
  // Standard attributes
  private _textAlignment: EditorV3Align = EditorV3Align.left;
  /**
   * Current text alignment
   */
  get textAlignment() {
    return this._textAlignment;
  }
  /**
   * Set text alignment
   * Can be left, center, right, or justify
   */
  set textAlignment(newTa: EditorV3Align) {
    if (newTa !== this._textAlignment) {
      this.lines.forEach((l) => (l.textAlignment = newTa));
      this._textAlignment = newTa;
    }
  }

  private _decimalAlignPercent = 60;
  /**
   * Current decimal alignment percent
   */
  get decimalAlignPercent() {
    return this._decimalAlignPercent;
  }
  set decimalAlignPercent(newDap: number) {
    if (newDap !== this._decimalAlignPercent) {
      this.lines.forEach((l) => (l.decimalAlignPercent = newDap));
      this._decimalAlignPercent = newDap;
    }
  }

  private _styles: EditorV3Styles = {};
  /**
   * Current available styles in the editor
   */
  get styles() {
    return this._styles;
  }
  set styles(newStyles: EditorV3Styles) {
    this._styles = newStyles;
  }

  private _markdownSettings = defaultMarkdownSettings;

  /**
   * Current lines in the editor
   */
  public lines: EditorV3Line[] = [];

  private _styleNode(): HTMLDivElement {
    const sn = document.createElement("div");
    sn.className = "aiev3-style-info";
    sn.dataset.style = JSON.stringify(this._styles);
    return sn;
  }

  // Read only attributes
  /**
   * Element to render
   */
  public toHtml(): DocumentFragment {
    const ret = new DocumentFragment();
    this.lines.forEach((l) => ret.append(l.toHtml()));
    if (Object.keys(this._styles).length > 0) ret.append(this._styleNode());
    return ret;
  }

  /**
   * Render Markdown in an Fragment
   * @param markdownSettings Markdown settings, current settings by default
   * @returns Documentfragment with rendered markdown
   */
  public toMarkdownHtml(
    markdownSettings: IMarkdownSettings = this._markdownSettings,
  ): DocumentFragment {
    const ret = new DocumentFragment();
    this.lines.forEach((line) => {
      const l = document.createElement("div");
      l.classList.add("aiev3-markdown-line");
      l.dataset.textAlignment = this._textAlignment;
      l.dataset.decimalAlignPercent = this._decimalAlignPercent.toString();
      const tn = document.createTextNode(line.toMarkdown(markdownSettings));
      l.append(tn);
      ret.append(l);
    });
    if (Object.keys(this._styles).length > 0) ret.append(this._styleNode());
    return ret;
  }

  /**
   * Text from the element
   */
  get text(): string {
    return this.lines.map((l) => l.textBlocks.map((b) => b.text).join("")).join("\n");
  }

  /**
   * JSON string of the content
   */
  get jsonString(): string {
    return JSON.stringify({
      lines: this.lines,
      textAlignment: this._textAlignment,
      decimalAlignPercent: this._decimalAlignPercent,
      styles: this._styles,
    });
  }

  /**
   * Current content properties
   */
  get contentProps(): EditorV3ContentProps {
    return {
      textAlignment: this._textAlignment,
      decimalAlignPercent: this._decimalAlignPercent,
      styles: this._styles,
    };
  }

  /**
   * Create a new EditorV3Content instance
   * @param input Stringified JSON or HTML/text
   * @param props Optional properties
   * @returns Instance of the object
   */
  constructor(input: string, props?: EditorV3ContentProps) {
    // Defaults
    this._textAlignment = props?.textAlignment ?? EditorV3Align.left;
    this._decimalAlignPercent = props?.decimalAlignPercent ?? 60;
    this._styles = props?.styles ?? {};
    this.lines = [];

    // Process incoming data
    try {
      // Check for stringified class input
      const jsonInput: EditorV3Import = JSON.parse(input);
      if (!Array.isArray(jsonInput.lines)) throw "No lines";
      this.copyImport(jsonInput, props);
    } catch {
      // Establish input as string
      const inputString = input as string;
      // Read in v3 HTML/text
      const r = readV3Html(inputString, this._textAlignment, this._decimalAlignPercent);
      this.copyImport(r, props);
      return;
    }
  }

  private copyImport(read: EditorV3Import, props?: EditorV3ContentProps): void {
    this._textAlignment =
      props?.textAlignment ?? (read.lines[0].textAlignment as EditorV3Align) ?? EditorV3Align.left;
    this._decimalAlignPercent =
      props?.decimalAlignPercent ?? read.lines[0].decimalAlignPercent ?? 60;
    this._styles = {
      ...this._styles,
      ...read.styles,
      ...props?.styles,
    };
    this.lines = read.lines.map((l) =>
      l instanceof EditorV3Line
        ? l
        : new EditorV3Line(JSON.stringify(l), this._textAlignment, this._decimalAlignPercent),
    );
  }

  public subLines(pos: EditorV3Position): EditorV3Line[] {
    const ret: EditorV3Line[] = [];
    if (
      pos.startLine === pos.endLine &&
      pos.startChar === pos.endChar &&
      pos.startLine < this.lines.length
    ) {
      const style = this.lines[pos.startLine].getStyleAt(pos.startChar);
      ret.push(
        new EditorV3Line(
          [new EditorV3TextBlock("", style)],
          this.textAlignment,
          this.decimalAlignPercent,
        ),
      );
    }
    // Check selection contains something
    if (
      pos.startLine < this.lines.length &&
      ((pos.startLine === pos.endLine && pos.startChar < pos.endChar) ||
        pos.startLine < pos.endLine)
    ) {
      // First line
      if (pos.startLine <= pos.endLine) {
        ret.push(
          new EditorV3Line(
            this.lines[pos.startLine].subBlocks(
              pos.startChar,
              pos.startLine === pos.endLine ? pos.endChar : undefined,
            ),
            this._textAlignment,
            this._decimalAlignPercent,
          ),
        );
      }
      // At least one line between
      if (pos.startLine + 1 < pos.endLine) {
        ret.push(...this.lines.slice(pos.startLine + 1, pos.endLine));
      }
      // End line to end character
      if (pos.startLine < pos.endLine && pos.endLine < this.lines.length) {
        ret.push(
          new EditorV3Line(
            this.lines[pos.endLine].upToPos(pos.endChar),
            this._textAlignment,
            this._decimalAlignPercent,
          ),
        );
      }
    }
    return ret;
  }

  public upToPos(endLine: number, endChar: number): EditorV3Line[] {
    return this.subLines({ startLine: 0, startChar: 0, endLine, endChar });
  }

  public fromPos(startLine: number, startChar: number): EditorV3Line[] {
    return this.subLines({
      startLine,
      startChar,
      endLine: this.lines.length - 1,
      endChar: this.lines[this.lines.length - 1].lineLength,
    });
  }

  public getStyleAt(line: number, character: number): string | undefined {
    return line < this.lines.length ? this.lines[line].getStyleAt(character) : undefined;
  }

  // Split line in two
  public splitLine(pos: EditorV3Position): EditorV3Position {
    const u = this.upToPos(pos.startLine, pos.startChar);
    const f = this.fromPos(pos.endLine, pos.endChar);
    const newLines = [...u, ...f];
    this.lines = newLines;
    return {
      isCollapsed: true,
      startLine: pos.startLine + 1,
      startChar: 0,
      endLine: pos.startLine + 1,
      endChar: 0,
    };
  }

  // Merge line with next line
  public mergeLines(line: number) {
    if (this.lines.length > line + 1) {
      this.lines[line].insertBlocks(this.lines[line + 1].textBlocks, this.lines[line].lineLength);
      this.lines.splice(line + 1, 1);
    }
  }

  public deleteCharacter(pos: EditorV3Position, backwards: boolean): EditorV3Position {
    if (this.lines.length > pos.startLine && this.text !== "") {
      const newPos = { ...pos };
      // Move cursor if start and end are the same
      if (pos.startLine === pos.endLine && pos.startChar === pos.endChar) {
        // Backspace at end of line
        if (backwards && pos.startLine > 0 && pos.startChar === 0) {
          newPos.startChar = this.lines[pos.startLine - 1].lineLength;
          newPos.startLine = pos.startLine - 1;
        }
        // Backspace after first character
        else if (backwards && pos.startChar > 0) {
          newPos.startChar = pos.startChar - 1;
        }
        // Delete end line
        else if (pos.endChar >= this.lines[pos.endLine].lineLength) {
          newPos.endChar = 0;
          newPos.endLine = pos.endLine + 1;
        } else if (pos.endChar < this.lines[pos.endLine].lineLength) {
          newPos.endChar = pos.endChar + 1;
        }
      }
      // Remove under selected
      this.splice(newPos);
      if (pos.isCollapsed) {
        newPos.endLine = newPos.startLine;
        newPos.endChar = newPos.startChar;
      }
      return newPos;
    }
    return pos;
  }

  /**
   * Splice, when converting everything to normal text
   * @param pos Position in markdown
   * @param newLines New lines (not markdown)
   * @returns Markdown cut as lines
   */
  private spliceMarkdown(pos: EditorV3Position, newLines?: EditorV3Line[]): EditorV3Line[] {
    const thisMarkdown = new EditorV3Content(
      this.lines.map((l) => l.toMarkdown(this._markdownSettings)).join("\n"),
    );
    const newLinesMarkdown = newLines?.map(
      (l) => new EditorV3Line(l.toMarkdown(this._markdownSettings)),
    );
    if (
      pos.startLine < thisMarkdown.lines.length &&
      pos.startLine <= pos.endLine &&
      !(pos.startLine === pos.endLine && pos.endChar < pos.startChar)
    ) {
      const ret: EditorV3Line[] = thisMarkdown.subLines(pos);
      const u = thisMarkdown.upToPos(pos.startLine, pos.startChar);
      const f = thisMarkdown.fromPos(pos.endLine, pos.endChar);
      thisMarkdown.lines = [...u, ...(newLinesMarkdown ?? []), ...f];
      // Change back from markdown
      this.lines = thisMarkdown.lines.map(
        (l) =>
          new EditorV3Line(
            new MarkdownLineClass({ line: l.lineText }).toTextBlocks(),
            this.textAlignment,
            this.decimalAlignPercent,
          ),
      );
      this.mergeLines(
        pos.startLine +
          (newLines?.length ?? 0) -
          (pos.startChar < this.lines[pos.startLine].lineLength ? 1 : 0),
      );
      newLines?.length && this.mergeLines(pos.startLine);
      return ret;
    }
    return [];
  }

  public splice(
    pos: EditorV3Position,
    newLines?: EditorV3Line[],
    asMarkdown = false,
  ): EditorV3Line[] {
    if (asMarkdown) {
      return this.spliceMarkdown(pos, newLines);
    } else if (
      pos.startLine < this.lines.length &&
      pos.startLine <= pos.endLine &&
      !(pos.startLine === pos.endLine && pos.endChar < pos.startChar)
    ) {
      const ret: EditorV3Line[] = this.subLines(pos);
      const u = this.upToPos(pos.startLine, pos.startChar);
      const f = this.fromPos(pos.endLine, pos.endChar);
      this.lines = [
        ...u,
        ...(newLines ? newLines.map((l) => new EditorV3Line(l.jsonString)) : []),
        ...f,
      ];
      this.mergeLines(
        pos.startLine +
          (newLines?.length ?? 0) -
          (pos.startChar < this.lines[pos.startLine].lineLength ? 1 : 0),
      );
      newLines?.length && this.mergeLines(pos.startLine);
      return ret;
    }
    return [];
  }

  public applyStyle(styleName: string, pos: EditorV3Position) {
    if (pos.startLine < this.lines.length && pos.startLine <= pos.endLine) {
      for (let _i = pos.startLine; _i <= pos.endLine && _i < this.lines.length; _i++) {
        this.lines[_i].applyStyle(
          styleName,
          _i === pos.startLine ? pos.startChar : 0,
          _i === pos.endLine ? pos.endChar : Infinity,
        );
      }
    }
    return this;
  }

  public removeStyle(pos: EditorV3Position) {
    if (pos.startLine < this.lines.length && pos.startLine <= pos.endLine) {
      for (let _i = pos.startLine; _i <= pos.endLine && _i < this.lines.length; _i++) {
        this.lines[_i].removeStyle(
          _i === pos.startLine ? pos.startChar : 0,
          _i === pos.endLine ? pos.endChar : Infinity,
        );
      }
    }
    return this;
  }
}
