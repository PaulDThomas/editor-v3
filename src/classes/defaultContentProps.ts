import { EditorV3Align, EditorV3ContentProps } from "./interface";
import { defaultMarkdownSettings } from "./markdown/MarkdownSettings";

export const defaultContentProps: EditorV3ContentProps = {
  allowMarkdown: false,
  allowNewLine: false,
  decimalAlignPercent: 60,
  markdownSettings: defaultMarkdownSettings,
  showMarkdown: false,
  styles: undefined,
  textAlignment: EditorV3Align.left,
  atListFunction: undefined,
  maxAtListLength: 10,
};
