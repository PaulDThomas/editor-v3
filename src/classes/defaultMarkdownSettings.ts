export interface IMarkdownSettings {
  styleStartTag: string;
  styleNameEndTag: string;
  styleEndTag: string;
  defaultStyle: string;
  dropDownStartTag: string;
  dropDownEndTag: string;
  dropDownValueSeparator: string;
  dropDownSelectedValueTag: string;
  calcStartTag: string;
  calcNameEndTag: string;
  calcEndTag: string;
  calcConditionSeparator: string;
  calcAndSeparator: string;
  atStartTag: string;
  atEndTag: string;
}

export const defaultMarkdownSettings = {
  styleStartTag: "<<",
  styleNameEndTag: "::",
  styleEndTag: ">>",
  defaultStyle: "defaultStyle",
  dropDownStartTag: "[[",
  dropDownEndTag: "]]",
  dropDownValueSeparator: "||",
  dropDownSelectedValueTag: "**",
  calcStartTag: "[![",
  calcNameEndTag: "::",
  calcEndTag: "]!]",
  calcConditionSeparator: "||",
  calcAndSeparator: "&&",
  atStartTag: "@[",
  atEndTag: "@]",
};
