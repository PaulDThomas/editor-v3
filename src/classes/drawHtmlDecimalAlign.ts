import { EditorV3BlockClass, EditorV3ContentProps, EditorV3RenderProps } from "./interface";

export const drawHtmlDecimalAlign = (
  renderProps: EditorV3RenderProps,
  contentProps: EditorV3ContentProps,
  lhsContent: EditorV3BlockClass[],
  rhsContent: EditorV3BlockClass[],
) => {
  if (renderProps.currentEl) {
    const prePoint = document.createElement("span");
    prePoint.className = "aiev3-span-point lhs";
    renderProps.currentEl.append(prePoint);
    if (lhsContent.length)
      lhsContent.forEach((tb) =>
        prePoint.append(
          tb.toHtml(
            { ...renderProps, currentEl: prePoint },
            tb.style ? contentProps.styles?.[tb.style] : undefined,
          ),
        ),
      );
    else prePoint.textContent = "\u200b";

    // Set up space after (excluding) decimal
    const postPoint = document.createElement("span");
    postPoint.className = "aiev3-span-point rhs";
    renderProps.currentEl.append(postPoint);
    if (rhsContent.length)
      rhsContent.forEach((tb) => {
        postPoint.append(
          tb.toHtml(
            { ...renderProps, currentEl: postPoint },
            tb.style ? contentProps.styles?.[tb.style] : undefined,
          ),
        );
      });
    else postPoint.textContent = "\u200b";

    // Set grid column widths
    renderProps.currentEl.style.gridTemplateColumns = `${contentProps.decimalAlignPercent}% ${100 - contentProps.decimalAlignPercent}%`;
  }
};
