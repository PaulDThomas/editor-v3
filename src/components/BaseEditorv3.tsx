import { useLayoutEffect, useState } from "react";
import baseStyles from "./BaseInputs.module.css";
import { EditorV3, EditorV3Props } from "./EditorV3";

interface BaseEdivotV3Props extends EditorV3Props {
  label?: string;
}

export const BaseEditorV3 = ({ label, style, resize, ...rest }: BaseEdivotV3Props): JSX.Element => {
  const [outerStyle, setOuterStyle] = useState<React.CSSProperties>(style ?? {});
  const [innerStyle, setInnerStyle] = useState<React.CSSProperties>(style ?? {});
  useLayoutEffect(() => {
    const recalc: React.CSSProperties = { ...style };
    const outer: React.CSSProperties = {};
    if (style) {
      if (style.width) {
        outer.width = style.width;
        recalc.width = "100%";
      }
      if (style.maxWidth) {
        outer.maxWidth = style.maxWidth;
        recalc.maxWidth = "100%";
      }
      if (style.minWidth) {
        outer.minWidth = style.minWidth;
        recalc.minWidth = "100%";
      }
      if (style.height) {
        outer.height = style.height;
        recalc.height = "100%";
      } else {
        outer.height = "48px";
        recalc.height = "100%";
      }
      if (style.maxHeight) {
        outer.maxHeight = style.maxHeight;
        recalc.maxHeight = "100%";
      }
      if (style.minHeight) {
        outer.minHeight = style.minHeight;
        recalc.minHeight = "100%";
      } else {
        outer.minHeight = "48px";
        recalc.minHeight = "100%";
      }
      if (resize === true || resize === "vertical") {
        outer.resize = "vertical";
        outer.overflow = "auto";
      }
    }
    setOuterStyle(outer);
    setInnerStyle(recalc);
  }, [style, resize]);
  return (
    <div
      className={baseStyles.holder}
      style={outerStyle}
    >
      {label !== undefined && (
        <label
          className={baseStyles.label}
          htmlFor={rest.id}
        >
          {label}
        </label>
      )}
      <div className={baseStyles.editorHolder}>
        <EditorV3
          {...rest}
          style={innerStyle}
          noBorder
          resize={false}
        />
      </div>
    </div>
  );
};
