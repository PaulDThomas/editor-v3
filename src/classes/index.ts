import { EditorV3Content } from "./EditorV3Content";
import { EditorV3Line } from "./EditorV3Line";
import { EditorV3AtBlock } from "./EditorV3AtBlock";
import { EditorV3TextBlock } from "./EditorV3TextBlock";
import { EditorV3Align, EditorV3Styles, IEditorV3, EditorV3DropListItem } from "./interface";

export { EditorV3Align, EditorV3Content, EditorV3Line, EditorV3TextBlock, EditorV3AtBlock };
export type { EditorV3Styles, IEditorV3, EditorV3DropListItem as EditorV3AtListItem };
