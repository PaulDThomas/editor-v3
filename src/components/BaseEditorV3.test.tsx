import { render, screen } from "@testing-library/react";
import { defaultMarkdownSettings } from "../classes/defaultMarkdownSettings";
import { EditorV3Content } from "../classes/EditorV3Content";
import { EditorV3Align } from "../classes/interface";
import { BaseEditorV3 } from "./BaseEditorv3";

const mockContent = new EditorV3Content("34.45\n\nx.xx", {
  allowMarkdown: true,
  allowNewLine: false,
  decimalAlignPercent: 80,
  markdownSettings: defaultMarkdownSettings,
  showMarkdown: false,
  styles: { shiny: { color: "pink", fontWeight: "700" } },
  textAlignment: EditorV3Align.decimal,
});
mockContent.applyStyle("shiny", { startLine: 2, startChar: 0, endLine: 2, endChar: 4 });

describe("BaseEditorV3", () => {
  test("Draw and fire cursor events", async () => {
    render(
      <div data-testid="container">
        <BaseEditorV3
          label="Base editor test"
          id="test-editor"
          input={mockContent}
          style={{
            minWidth: "200px",
            width: "600px",
            maxWidth: "800px",
            minHeight: "100px",
            height: "200px",
            maxHeight: "400px",
          }}
          customStyleMap={{ shiny: { color: "pink", fontWeight: "700" } }}
          allowNewLine
          resize
        />
      </div>,
    );
    const container = (await screen.findByTestId("container")).children[0] as HTMLDivElement;
    expect(screen.queryByText("Base editor test")).toBeInTheDocument();
    expect(screen.queryByText("Base editor test")).toBeInTheDocument();
    expect(container.outerHTML).toMatchSnapshot();
  });
});
