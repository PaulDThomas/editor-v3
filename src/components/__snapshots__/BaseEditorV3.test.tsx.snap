// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseEditorV3 Basic render 1`] = `"<div><div class="holder" style="height: calc(1.5em + 26px); min-height: calc(1.5em + 26px);"><label class="label" id="test-editor-label" for="test-editor">Base editor test</label><div class="editorHolder"><div aria-labelledby="test-editor-label" class="aiev3" id="test-editor"><div style="width: 100%; height: 100%;" class="aiw-ContextMenu-module-WinFxq-contextMenuHandler "><div class="aiev3-resize" style="background-color: lightgray; height: 100%; min-height: 100%;"><div id="test-editor-editable" class="aiev3-editing singleline" contenteditable="true" role="textbox" spellcheck="false"><div class="aiev3-line left"><span class="aiev3-tb" data-line-start-position="0"> </span></div></div></div></div></div></div></div></div>"`;

exports[`BaseEditorV3 Draw and fire cursor events 1`] = `"<div class="holder" style="min-width: 200px; width: 600px; max-width: 800px; min-height: 100px; height: 200px; max-height: 400px; resize: vertical; overflow: auto;"><label class="label" id="test-editor-label" for="test-editor">Base editor test</label><div class="editorHolder"><div aria-labelledby="test-editor-label" class="aiev3" id="test-editor"><div style="width: 100%; height: 100%;" class="aiw-ContextMenu-module-WinFxq-contextMenuHandler "><div class="aiev3-resize" style="min-width: 100%; width: 100%; max-width: 100%; min-height: 100%; height: 100%; max-height: 100%;"><div id="test-editor-editable" class="aiev3-editing multiline" contenteditable="false" spellcheck="false"><div class="aiev3-line left"><span class="aiev3-tb" data-line-start-position="0">34.45</span></div><div class="aiev3-line left"><span class="aiev3-tb" data-line-start-position="0"> </span></div><div class="aiev3-line left"><span class="aiev3-tb editorv3style-shiny" data-line-start-position="0" data-style-name="shiny" style="color: pink; font-weight: 700;">x.xx</span></div></div></div></div></div></div></div>"`;
